snuba/consumers/consumer_builder.py

```diff
@@ -109,29 +129,42 @@ def __init__(
 
         stream_loader = self.storage.get_table_writer().get_stream_loader()
 
-        self.raw_topic: Topic
+        self.raw_topic: ArroyoTopic
         if kafka_params.raw_topic is not None:
-            self.raw_topic = Topic(kafka_params.raw_topic)
+            # we do not support topic override in a sliced context
+            assert slice_id is None
```

Comment:
Why not?

Reply:
This was part of a discussion I had with Oliver, and he'd said that we do not plan to support this as of now.

Reply:
Since we already support topic override on unsliced storages, the effort to do so on a sliced storage which uses the same code seems pretty small. Almost the same as throwing this AssertionError here.

I don't think it's a good idea to not support this as there is no particular reason for it.

In general I think we should not be treating sliced storages as a special case unless there is a very strong reason for omitting a feature.

Reply:
Sure, this makes sense to me. I can have the topic override exist for sliced settings as well.

Reply:
Is topic override for sliced environments a hard requirement? I know it is good to have. But we are building iteratively and feel like this can be supported at a later point in time.

Reply:
Isn't the effort to support a sliced storage literally just removing the assertion on this one line? Not sure why we would want to not do that now and add it later.

Reply:
Under what circumstances do we use the topic overrides? From my understanding, it is for overriding topics which are hard-coded in the code based on different environments. With slicing, there is an explicit way of overriding topics per slice which can be seen here https://github.com/getsentry/ops/blob/master/k8s/services/snuba/templates/prod_settings.py#L469-L496. This keeps the method of overrides consistent across different model objects(kafka topics, kafka brokers, clusters, etc.) which can be sliced. So the precedence order in a sliced environment is always prefer slice configuration if it exists.

Now individual model objects can provide their own mechanism of overrides. In a non-sliced environment that works fine. But if we start allowing overrides for different model objects and the precedence order of the different model objects varies by type of model, it makes things complicated.

Reply:
What you described about the topic overrides for slicing is not any different for non sliced topics. There also exists an explicit way to override the topics in settings. It's essentially exactly the same - KAFKA_TOPIC_MAP and SLICED_KAFKA_TOPIC_MAP. Slicing and non slicing is exactly the same in this regard, the value just gets picked up from a different variable in settings.

There are scenarios where we want to pass via the CLI. Doing the errors/transactions topic split was a recent example where it was needed. The precedence (CLI always takes precedence!) should be exactly the same as for non sliced topics and it's like this on all our consumers, replacers, etc. I don't think there is anything unique or extra complicated when there is a slice involved.

If you look a few lines below this one you will see that this PR already supports overrides for the replacement and commit log topics via CLI for both sliced and non sliced consumer. So I don't really get why we woudn't also want the same for the raw topic.

Reply:
You bring up a good point. The CLI can override topic names today. The question I would like to answer is should we allow CLI overrides when slice_id is passed? There can be conflicts between the topic-name passed in via CLI and the configuration in SLICED_KAFKA_TOPIC_MAP.

Lets assume you have settings like these
SLICED_KAFKA_TOPIC_MAP -> (orig-topic-name, 1) = new-topic-name-1
The consumer code is run with 
`consumer --slice-id=1 --topic=cli-topic-name ...`
When the consumer tries to resolve the topic name, it tries to lookup using `cli-topic-name` instead of `orig-topic-name`. The lookup would fail. The question is how do we avoid these human errors. The problem of trying to keep 2 different sets of configuration in sync is a manual operation.

A simpler way to avoid this problem might be to disallow topic overrides when slice-id is passed. The current usage of topic overrides can be achieved by modifying the configuration of `SLICED_KAFKA_TOPIC_MAP` itself. I am open to other suggestions as well.

Reply:
I disagree that just disallowing topic overrides is a simple way of solving the problem. I think that it makes it more complex because what it does is make sliced logic and unsliced logic different in this code path, when this code path should not be concerned with slicing at all and should behave exactly the same way in both scenarios.

About your comment about lookups failing. I believe it's because of a separate bug in how resolving is done in this PR. Left a comment on it here https://github.com/getsentry/snuba/pull/3259#discussion_r1014501873. If this is fixed we can just remove this one line and everything behaves the same in both sliced and unsliced variants.

Reply:
I think I follow Lyn's suggestion, and indeed using the default topic spec can allow us to resolve consistently and accurately to the physical topic name. It should be a small change, so if we can easily retain the original behavior of the ConsumerBuilder, I think it makes sense to do so.